agdk/agdktunnel/app/src/main/cpp/vibration_manager.cpp

```diff
@@ -0,0 +1,109 @@
+/*
+ * Copyright 2023 The Android Open Source Project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "vibration_manager.hpp"
+#include "android/content/context.hpp"
+#include "android/os/vibrator_manager.hpp"
+#include <android/api-level.h>
+#include <gni/gni.h>
+#include <jni.h>
+
+// Add a VibrationAttribute to the vibrate call on API 33 or higher
+#define USE_VIBRATION_ATTRIBUTE_API_LEVEL 33
+
+// Use VibrationManager to obtain a vibrator on API 31 and higher
+#define USE_VIBRATION_MANAGER_API_LEVEL 31
+
+// Use the VibrationEffect version of the vibrate call on API 26 or higher
+#define USE_VIBRATION_EFFECT_API_LEVEL 26
+
+// android.os.VibrationAttributes.USAGE_CLASS_FEEDBACK
+#define USAGE_CLASS_FEEDBACK 2
+
+// android.os.VibrationEffect.EFFECT_HEAVY_CLICK
+#define EFFECT_HEAVY_CLICK 5
+
+// Amount of time to vibrate in milliseconds using the legacy vibrate method
+#define VIBRATION_TIME_MILLISECONDS 500
```

Comment:
The heavy click is actually quite shorter than this. I'd suggest something like 30ms for the "fallback" vibration here.